primitives/src/unique_vec.rs

```diff
@@ -0,0 +1,334 @@
+    index: usize,
+}
+
+/// Wrapper type for [`Vec`] which ensures that all elements are unique.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Encode, IntoSchema)]
+pub struct UniqueVec<T: PartialEq>(Vec<T>);
+
+impl<T: PartialEq> Default for UniqueVec<T> {
+    fn default() -> Self {
+        Self(Vec::default())
+    }
+}
+
+impl<T: PartialEq> UniqueVec<T> {
+    /// Create new [`UniqueVec`].
+    #[must_use]
+    pub const fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Try construct [`UniqueVec`] from iterator.
+    ///
+    /// # Errors
+    ///
+    /// - If there are duplicates in the sequence.
+    pub fn try_from_iter<I>(iter: I) -> Result<Self, DuplicationError>
+    where
+        I: IntoIterator<Item = T>,
+    {
+        let mut v = UniqueVec::new();
+        v.try_extend(iter)?;
+        Ok(v)
+    }
+
+    /// Try extend [`UniqueVec`] from iterator.
+    /// If an error occurs, all previously added elements will remain in the [`UniqueVec`].
+    ///
+    /// # Errors
+    ///
+    /// - If there are duplicates in the sequence.
+    pub fn try_extend<I>(&mut self, iter: I) -> Result<(), DuplicationError>
+    where
+        I: IntoIterator<Item = T>,
+    {
+        for (i, value) in iter.into_iter().enumerate() {
+            if !self.push(value) {
+                return Err(DuplicationError { index: i });
+            }
+        }
+        Ok(())
+    }
+
+    /// Push `value` to [`UniqueVec`] if it is not already present.
+    ///
+    /// Returns `true` if value was pushed and `false` if not.
+    pub fn push(&mut self, value: T) -> bool {
+        if self.contains(&value) {
+            false
+        } else {
+            self.0.push(value);
+            true
+        }
+    }
+
+    /// Removes the element at the given `index` and returns it.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `index` is out of bounds.
+    pub fn remove(&mut self, index: usize) -> T {
+        self.0.remove(index)
+    }
+
+    /// Clears the [`UniqueVec`], removing all values.
+    pub fn clear(&mut self) {
+        self.0.clear()
+    }
+
+    /// Returns reference to an internal mutable slice.
+    ///
+    /// # Warning
+    ///
+    /// Caller is responsible for ensuring that elements in slice will remain unique.
+    pub fn as_mut_unchecked(&mut self) -> &mut [T] {
+        &mut self.0
+    }
+}
+
+impl<T: PartialEq> TryFrom<Vec<T>> for UniqueVec<T> {
+    type Error = DuplicationError;
+
+    fn try_from(value: Vec<T>) -> Result<Self, Self::Error> {
+        let mut inner = Vec::with_capacity(value.len());
+        for (i, item) in value.into_iter().enumerate() {
+            if inner.contains(&item) {
+                return Err(DuplicationError { index: i });
+            }
+            inner.push(item);
+        }
+        Ok(Self(inner))
+    }
+}
+
+impl<T: PartialEq> From<UniqueVec<T>> for Vec<T> {
+    fn from(value: UniqueVec<T>) -> Self {
+        value.0
+    }
+}
```

Comment:
IMO implementing `FromIterator` is sufficient. I would remove these

Reply:
I cannot return `Err` from `FromIterator`

Reply:
yes, you can. You have to make it like this:

```rs
impl<T: PartialEq> FromIterator<UniqueVec<T>> for Result<Vec<T>, DuplicationError>
```

Reply:
Em... The thing you wrote works like `flat()`, I think you meant:

```rust
impl<T: PartialEq> FromIterator<T> for Result<UniqueVec<T>, DuplicationError>
```

Reply:
I think you meant something different, because the thing you wrote works like `flatten()`.

Maybe you meant:

```rust
impl<T: PartialEq> FromIterator<T> for Result<UniqueVec<T>, DuplicationError>
```

which won't work, see [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fadcc2c9b71498fe0ab7355c1336cc4d)